treesitter config: query_file_ignore

MunifTanjim opened the pull request:
resolves #15244

## Description

- Introduces `vim.treesitter.config` method. It is used for updating configuration options for treesitter.
- Introduces treesitter configuration option: `query_file_ignore`

## Usage

### Usage: `vim.treesitter.config(options, scope)`

Treesitter config can be set for these:
- for any language
- for a specific language
- for a specific buffer

**set config for a specific buffer**

```lua
vim.treesitter.config({
  query_file_ignore = {},
}, 0)
```

**set config for a specific language**

```lua
vim.treesitter.config({
  query_file_ignore = {},
}, "lua")
```

**set config for any language**

```lua
vim.treesitter.config({
  query_file_ignore = {},
})
```

### Usage: config `query_file_ignore`

_Lua string pattern: https://www.lua.org/pil/20.2.html_

**Ignore all query files from a specific plugin**:

```lua
vim.treesitter.config({
  query_file_ignore = { "/nvim%-treesitter/" },
}, "lua")
```

**Ignore specific query file from a specific plugin**:

```lua
vim.treesitter.config({
  query_file_ignore = { "/nvim%-treesitter/.+/indents.scm" },
}, "lua")
```

**Ignore all query files from all plugins**:

```lua
vim.treesitter.config({
  query_file_ignore = { ".+" },
}, "lua")
```

**Add/Remove a specific ignore file pattern**:

```lua
vim.treesitter.config({
  query_file_ignore = function(patterns)
    table.insert(patterns, "/nvim%-treesitter/") -- or filter it out to remove
    return patterns
  end,
}, "lua")
```

**Clear all ignore file patterns**:

```lua
vim.treesitter.config({
  query_file_ignore = {},
}, "lua")
-- or
vim.treesitter.config({
  query_file_ignore = function(patterns)
    return nil
  end,
}, "lua")
```

MunifTanjim committed:
feat(treesitter): add vim.treesitter.config

gpanders commented:
Is there a way to re-enable a query file after it's been ignored? I can see that being useful for e.g. debugging or to allow easy toggling of certain query files.

vigoux commented on runtime/lua/vim/treesitter/query.lua:
```diff
@@ -34,14 +69,29 @@ local function safe_read(filename, read_quantifier)
+---@param lang string
+---@param patterns string[]
+function M.ignore_query_files(lang, patterns)
+  if not query_file_config.ignore_patterns[lang] then
+    query_file_config.ignore_patterns[lang] = {}
+    query_file_config.ignored_files[lang] = {}
+  end
```
Following @gpanders comment, I would do the following instead:
```suggestion
  query_file_config.ignore_patterns[lang] = {}
  query_file_config.ignored_files[lang] = {}
```

MunifTanjim commented:
> Is there a way to re-enable a query file after it's been ignored? I can see that being useful for e.g. debugging or to allow easy toggling of certain query files.

We can go with one of these three alternatives.

---

```lua
set_query_ignore_patterns(lang: string, patterns: string[] | fn | nil)
```

If `patterns` is array of string `string[]`, set it as ignore patterns.

If `patterns` is function `fn`, it takes the current ignore patterns and returns the next ignore patterns. So user can add more patterns to it or remove some.

**Pros**:
- Flexible API
- Can be called multiple times (from different places)

**Cons**:
- Might not seem so friendly for some people (?)

---

Or, same functionality as the above example, but more explicit API:

```lua
set_query_ignore_patterns(lang: string, patterns: string[], action?: 'replace' | 'add' | 'remove' = 'replace')
```

Based on the value of `action` (defaults to 'replace'), `patterns` will be added to the current list, removed from it or replace it.

**Props**:
- More verbose than the previous example (can go under cons too, depends on the people)
- Flexible API
- Can be called multiple times (from different places)


**Cons**:
- No way to read the current value of ignore patterns.

---

Or go with a restricted set API that always replaces the current list:

```lua
set_query_ignore_patterns(lang: string, patterns: string[] | nil)
```

**Pros**
- Simplest implementation

**Cons**
- Since it always replaces the value, calling multiple times (from different places) might break things.
- No way to read the current value of ignore patterns.
 
---

@vigoux you made a comment earlier (now deleted) which suggested the third example here.

@gpanders 

Now that the pros and cons are laid out, what do you guys think we should go with?

Or maybe there are some even simpler ways to acheive this?

vigoux commented on runtime/lua/vim/treesitter/query.lua:
```diff
@@ -34,14 +72,45 @@ local function safe_read(filename, read_quantifier)
   return content
 end
 
+-- Disables loading query files for {lang} matching provided {patterns}.
+--
+-- If {patterns} is nil, it clears the ignore patterns.
+-- If {patterns} is string[], it replaces the currently set patterns (if any).
+--
+-- You can also pass a function as {patterns}, which will accept a table with
+-- currently set patterns as keys. You can modify that table and return a new
+-- table or nil. The return value will be used as the new ignore patterns.
+--
+---@param lang string
+---@param patterns nil | string[] | fun(patterns:table<string, boolean | nil>): nil | table<string, boolean | nil>
```
We use `---` instead of `--` for doc comments:
```suggestion
--- Disables loading query files for {lang} matching provided {patterns}.
---
--- If {patterns} is nil, it clears the ignore patterns.
--- If {patterns} is string[], it replaces the currently set patterns (if any).
---
--- You can also pass a function as {patterns}, which will accept a table with
--- currently set patterns as keys. You can modify that table and return a new
--- table or nil. The return value will be used as the new ignore patterns.
---
---@param lang string
---@param patterns nil | string[] | fun(patterns:table<string, boolean | nil>): nil | table<string, boolean | nil>
```

vigoux commented on runtime/lua/vim/treesitter/query.lua:
```diff
@@ -34,14 +72,45 @@ local function safe_read(filename, read_quantifier)
+    return
+  end
+
+  query_file_config.ignore_patterns[lang] = {}
+  for _, pattern in ipairs(patterns) do
+    query_file_config.ignore_patterns[lang][pattern] = true
+  end
```
Why not using a table directly here ? This is not the classic way of storing value here :wink:

MunifTanjim replied:
I did that to support this pattern:

```lua
vim.treesitter.ignore_query_files("lua", function(patterns)
  patterns["/nvim%-treesitter/"] = true -- or false/nil for removing
  return patterns
end)
```

I thought it might be handy when somebody wants to toggle a pattern on/off (via some keybinding or commands). Otherwise they have to loop over the table (`string[]`) and filter out the value... or check if the value already exists before adding the value.

Do you want store the table (as `string[]`) directy instead of making `table<string, boolean>`?

vigoux replied:
To be honest yes, I'd rather keep it simple.
The simplest it is, the better.

MunifTanjim replied:
Updated 👍🏼

vigoux commented:
Could you add a little test for this feature ? After that we'll be good to go !

MunifTanjim commented:
> Could you add a little test for this feature ? After that we'll be good to go !

Added the tests for this method. Since we only have a single query file to play with `runtime/queries/c/highlights.scm`, that's all I could think of testing... hoping it would be enough 🤞🏼

vigoux reviewed (commented):
You are going to hate me but : could you run `./scripts/gen_vimdoc.py -t treesitter` just to see the look of the generated docs ? I think we handle type information in lua pretty bad.

MunifTanjim commented:
> You are going to hate me but : could you run ./scripts/gen_vimdoc.py -t treesitter just to see the look of the generated docs ? I think we handle type information in lua pretty bad.

😂 no worries... This is what I get:

![image](https://user-images.githubusercontent.com/8050659/129165470-fd3a0e07-a769-494f-a6cb-0f07fd47d94b.png)

Should I remove the type information?

tjdevries commented:
Whatever makes the docs look good for now is fine with me :)

MunifTanjim commented:
I've removed the type annotations.

justinmk commented on runtime/lua/vim/treesitter/query.lua:
```diff
@@ -6,17 +6,55 @@ local language = require'vim.treesitter.language'
+---@param file string
+---@return boolean
+local function is_file_ignored(lang, file)
+  local ignore_patterns = query_file_config.ignore_patterns[lang]
```
if we are going to create an inner configuration platform for treesitter (rather than using e.g. Nvim editor options, `option.c`) can we at least not add random functions to `vim.treesitter` and instead have a single `vim.treesitter.config` interface or something like that?

justinmk commented on runtime/lua/vim/treesitter/query.lua:
```diff
@@ -34,14 +72,36 @@ local function safe_read(filename, read_quantifier)
+---
+---@param lang The language
+---@param patterns Ignore patterns or callback function
+function M.ignore_query_files(lang, patterns)
```
why is this in treesitter/query.lua instead of toplevel `treesitter.lua`? I am sure this is only the start of the configuration requests, and I could imagine more configuration related to languages, highlighting--not only queries.

MunifTanjim replied:
Added this here because the `set_query` function was in this file. I'm open to changing it, just point me towards where it should go and how it should be implemented. 😃

justinmk commented:
- this adds a specialized function for a single purpose, `ignore_query_files`. How does one _undo_ this action?
- is there a conceivable need for this to be buffer-specific? then the current `ignore_query_files(lang, pattern)` signature is not enough.
- does it make sense to have something like `vim.treesitter.config(lang, bufnr, config, behavior)`
    - if `bufnr` is nil the configuration is applied to all buffers
    - `config` is a map of various settings, which is merged according the value of `behavior` (which acts like `vim.tbl_extend()`)

MunifTanjim commented:
> is there a conceivable need for this to be buffer-specific? then the current ignore_query_files(lang, pattern) signature is not enough.

Unless Neovim allows using multiple parser for same language, I don't think it's needed (but then again, I don't know if Neovim allows that)

> - does it make sense to have something like vim.treesitter.config(lang, bufnr, config, behavior)
>   - if bufnr is nil the configuration is applied to all buffers
>   - config is a map of various settings, which is merged according the value of behavior (which acts like vim.tbl_extend())
> - alterative: introduce a global-local Nvim option 'tsconfig' that defines the treesitter configuration.

As long as it lets the users be in control,  I'm open to both. I think Neovim's maintainers should be the person to decide which interface will be more maintainable in the long run. So, what's your suggestion? @justinmk 😃

justinmk commented:
> does it make sense to have something like vim.treesitter.opt(bufnr, lang, config, behavior)

This is my suggestion. (Renamed to "opt" instead of "config", for parallel with `vim.opt`)

MunifTanjim commented:
@justinmk

> vim.treesitter.opt(lang, bufnr, config, behavior)

Can you please give a hint about how to do buffer specific config without the table growing indefinitely (i.e. how to do cleanup)?

Also I noticed that `get_query_files(lang, query_name, is_included)` function doesn't accept `bufnr`. So in this case, the config for ignoring query files won't need `bufnr`. Maybe we can skip implementing the buffer specific config for now?

> config is a map of various settings, which is merged according the value of behavior (which acts like vim.tbl_extend())

This would mean the user won't be able to read the current config, right? But reading that would be necessary.

So should we expose the config table via `vim.treesitter.o` or `require('vim.treesitter.options')` or something?

In case we do that, do we need the `vim.treesitter.opt(bufnr, lang, config, behavior)` anymore? User can modify that table directly, right? 🤔

justinmk commented:
> Maybe we can skip implementing the buffer specific config for now?

Sure, as long as the API we choose now leaves the door open, e.g. by using "kwargs" pattern (`function foo({a=...,b=...})` instead of `function foo(a, b)`).

> Can you please give a hint about how to do buffer specific config without the table growing indefinitely (i.e. how to do cleanup)?

I would consider storing it as a buffer-local variable (note: this is serialized)

    :lua vim.b._nvim_treesitter_config = {foo='bar'}

MunifTanjim commented:
> Sure, as long as the API we choose now leaves the door open

Update the signature to this: `vim.treesitter.config(options, lang, bufnr)` (bufnr is optional)

> by using "kwargs" pattern (function foo({a=...,b=...}) instead of function foo(a, b)).

Let me know if you want kwargs pattern `{ config, lang, bufnr }` or something else.

justinmk commented:
In https://github.com/neovim/neovim/pull/16008 we are now considering a 'treesitteropt' editor builtin option. Please comment there if you have a preference on that vs `vim.treesitter.config()`.

MunifTanjim commented:
> In #16008 we are now considering a 'treesitteropt' editor builtin option. Please comment there if you have a preference on that vs vim.treesitter.config().

Just curious, what can we do to make the decision between `vim.bo.treesitteropt` vs `vim.treesitter.config(...)`?

I'm now biased towards `vim.treesitter.config(...)`, since we already have similar `vim.diagnostic.config(...)`. But I might not be seeing the bigger picture here.

Would it be possible to get opinions from a few more core maintainers?

@justinmk

clason commented:
@neovim/treesitter @gpanders

clason commented:
I think `vim.treesitter.config()` should
1. be reserved for _global_ options and
2. hence take only a single table.

Anything language or buffer specific is better handled through explicit `vim.treesitter.set_query` (or somesuch) functions.

MunifTanjim commented:
> Anything language or buffer specific is better handled through explicit `vim.treesitter.set_query` (or somesuch) functions;

I actually went with `vim.treesitter.ignore_query_files` initially, but in the previous discussions (in this PR) @justinmk suggested the `vim.treesitter.config` approach. Should we revert to that?

> maybe the best approach is extending `vim.treesitter.require_language` to take an optional dict with an allow list and/or a block list?

`vim.treesitter.require_language` seems to deal with only the parser, not queries. Might not be the right place for it? 🤔 

> Note that for `vim.treesitter.highlighter`, we already allow passing a {queries = {...}} option to override the queries used for that parsed tree.

This is not only about highlighting. The use-case goes like this: "I use a plugin that implements a feature using treesitter. They also provide some queries in `queries/<lang>/` folder. But I want to completely ignore those queries for a specific `<lang>` and write my own queries for it.

clason commented:
> I actually went with `vim.treesitter.ignore_query_files` initially, but in the previous discussions (in this PR) @justinmk suggested the `vim.treesitter.config` approach. Should we revert to that?

No, we should first discuss what kind of API we want to have before implementing anything.

> `vim.treesitter.require_language` seems to deal with only the parser, not queries. Might not be the right place for it? 🤔

Maybe not (but the parser implies which queries are loaded currently, so it's not unrelated); that's what needs to be discussed.

> This is not only about highlighting. The use-case goes like this: "I use a plugin that implements a feature using treesitter. They also provide some queries in `queries/<lang>/` folder for that feature. But I want to completely ignore those queries for a specific `<lang>` and write my own queries for it."

I know; I'm just saying there is already precedent for customizing the queries, and we should keep that pattern in mind. In particular, the use case you describe is _not_ about ignoring (specific) queries but _overriding_ them completely with your custom list. Basically, what your plugin does is implement something on the same level as `vim.treesitter.highlight`, so it's reasonable to follow a similar pattern.

(In particular, I don't quite see why in this case your plugin couldn't go straight for custom queries `<foo>.scm`, similarly to `indent.scm` and `fold.scm`? It's also important to keep separate Neovim's core treesitter API and the nvim-treesitter infrastructure on top of that.)

clason commented:
Let me put it this way: I worry that your proposal acts on the wrong level: queries belong to "modules" (in nvim-treesitter terminology), not to parsers and even less to the global scope. Your use case is valid, but I don't think a global ignore list is the right approach to handle it -- we want finer-grained control.

MunifTanjim commented:
> In particular, I don't quite see why in this case your plugin couldn't go straight for custom queries `<foo>.scm`, similarly to `indent.scm` and `fold.scm`?

Hmm... Sorry, the example of the use-case in my previous comment was probably bad. Let me take a step back and describe why I wanted this to begin with:

I was not satisfied with the default lua parser `nvim-treesitter` comes with (it's not really maintained and was incomplete). And so I wrote one. Now, all the plugin that provides queries for `lua` is for that default lua parser from `nvim-treesitter`. So for using this new parser there are two options:

1. Name it something else than `lua` (`lua-new` or whatever) so that I can put queries in `queries/lua-new` folder. And use that parser for `lua` language.
2. Put _something_ in my personal config file that tells neovim to ignore the queries files in `queries/lua` directories in all those plugins that comes with it. And only load queries for `lua` from my own `queries/lua` directory (which includes `indent.scm`, `folds.scm`, `textobjects.scm` and so on).

This PR is for implementing that "something" from Option 2 which I can put in my personal config, e.g.

```lua
  vim.treesitter.config({
    query_file_ignore = {
      "nvim-treesitter",
      "nvim-treesitter-textobjects",
      "nvim-treesitter-textsubjects",
      "refactoring.nvim",
    },
  }, "lua")
```

I think now we're more clear about the use-case I had.

With that being said, I understand if this is something that should not be included in Neovim core or if it's not worth adding the complexity. In that case I'll gladly go with the Option 1.

clason commented:
OK, that is indeed a very different problem. Yes, in general option 1 is the correct thing to do (different, incompatible, parsers should not have the same name!); but the real issue here is not with the core tree-sitter API but with nvim-treesitter (installing parsers) -- so I'd recommend simply _not_ installing the default parser and adding yours as a custom parser (as explained in the README).

In this case, however, if your Lua parser is indeed better than the original one by @tjdevries, we might wish to switch -- but that is again something handled by nvim-treesitter, so I'd suggest opening an issue there about switching.

MunifTanjim commented:
My use-case is unblocked by doing it a bit differently. Closing this PR.

Thanks everybody for the discussion! 💜

theHamsta commented:
@MunifTanjim I think this problem is not solved. We should have a way to select the query files

MunifTanjim commented:
@theHamsta Once the preferred solution is decided in https://github.com/nvim-treesitter/nvim-treesitter/issues/2241, I can open another PR if needed.

clason commented:
I'd like to keep the discussion open, even if the specific code here is probably not going to be merged. The generic question of how to override queries is still valid. (It could also be made into an issue, but for now it's easier to keep the history here.)

justinmk commented:
https://github.com/neovim/neovim/issues/15244 was closed so I assume this PR is abandoned.
